opt/scopeserver/dataserver/modules/gserver/GServer.py

```diff
@@ -580,19 +580,57 @@ def getMarkerGenes(self, request, context):
             md_cmm = md_clustering["clusterMarkerMetrics"]
 
             def create_cluster_marker_metric(metric):
-                cluster_marker_metrics = loom.get_cluster_marker_metrics(
-                    clustering_id=request.clusteringID, cluster_id=request.clusterID, metric_accessor=metric["accessor"]
+                return loom.get_cluster_marker_metrics(
+                    clustering_id=request.clusteringID,
+                    cluster_id=request.clusterID,
+                    metric_accessor=metric["accessor"]
                 )
+
+            def to_pb(metric):
                 return s_pb2.MarkerGenesMetric(
                     accessor=metric["accessor"],
                     name=metric["name"],
                     description=metric["description"],
-                    values=cluster_marker_metrics,
+                    values=cluster_marker_metrics[metric["accessor"]]
                 )
 
-            cluster_marker_metrics = list(map(create_cluster_marker_metric, md_cmm))
+            cluster_marker_metrics = functools.reduce(
+                lambda left, right: pd.merge(
+                    left,
+                    right,
+                    left_index=True,
+                    right_index=True,
+                    how='outer'
+                ),
+                list(
+                    map(
+                        lambda x: create_cluster_marker_metric(x),
+                        md_cmm
+                    )
+                )
+            )
+            # Keep only non-zeros elements
+            nonzero_mask = cluster_marker_metrics.apply(
+                lambda x: functools.reduce(
+                    np.logical_and,
+                    np.logical_and(
+                        ~np.isnan(x),
+                        x != 0
+                    )
+                ),
+                axis=1
+            )
+            cluster_marker_metrics_nonzero = cluster_marker_metrics[nonzero_mask]
 
-        return s_pb2.MarkerGenesReply(genes=genes, metrics=cluster_marker_metrics)
+        return s_pb2.MarkerGenesReply(
+            genes=cluster_marker_metrics_nonzero.index,
+            metrics=list(
+                map(
+                    to_pb,
+                    md_cmm
+                )
+            )
+        )
```

Comment:
I would prefer you assign this an identifier before passing it to `MarkerGenesReply`. `map` (sadly) is generally frowned upon in Python.
```suggestion
        metrics = [to_pb(x) for x in md_cmm]
        return s_pb2.MarkerGenesReply(
            genes=cluster_marker_metrics_nonzero.index,
            metrics=metrics
        )
```

Reply:
I'm happy to hear best practices in Python, also eager to know why ? :)

Reply:
A blog post by Guido is a good start: https://www.artima.com/weblogs/viewpost.jsp?thread=98196
Also top comment by Guido on PEP 279: https://www.python.org/dev/peps/pep-0279/

Reply:
Great! thanks for the links :+1: